hoomd/md/AlchemicalPotentialPair.h

```diff
@@ -0,0 +1,385 @@
+                    ArrayHandle<Scalar>(particle->m_alchemical_derivatives));
+                }
+            else
+                {
+                pkg.force_handles.push_back(ArrayHandle<Scalar>(GlobalArray<Scalar>()));
+                }
+        }
+    return pkg;
+    }
+
+template<class evaluator>
+inline void AlchemicalPotentialPair<evaluator>::pkgPerNeighbor(const unsigned int& i,
+                                                               const unsigned int& j,
+                                                               const unsigned int& typei,
+                                                               const unsigned int& typej,
+                                                               const bool& in_rcut,
+                                                               evaluator& eval,
+                                                               AlchemyPackage<evaluator>& pkg)
+    {
+    unsigned int alchemy_index = m_alchemy_index(typei, typej);
+    mask_type& mask {pkg.compute_mask[alchemy_index]};
+    alpha_array_t& alphas {pkg.alphas[alchemy_index]};
+    if (m_normalized)
+        eval.setNormalizationValue(pkg.normalizationValues[alchemy_index]);
+
+    // TODO: make sure that when we disable an alchemical particle, we rewrite its parameter
+
+    // calculate alchemical derivatives if needed
+    if (pkg.calculate_derivatives && in_rcut && mask.any())
+        {
+        std::array<Scalar, evaluator::num_alchemical_parameters> alchemical_derivatives = {};
+        eval.evalAlchDerivatives(alchemical_derivatives, alphas);
+        for (unsigned int k = 0; k < evaluator::num_alchemical_parameters; k++)
+            {
+            if (mask[k])
+                {
+                pkg.force_handles[alchemy_index].data[i]
+                    += alchemical_derivatives[k] * Scalar(-0.5);
+                pkg.force_handles[alchemy_index].data[j]
+                    += alchemical_derivatives[k] * Scalar(-0.5);
+                }
+            alchemy_index += m_alchemy_index.getNumElements();
+            }
+        }
+
+    // update parameter values with current alphas (MUST! be performed after dAlpha calculations)
+    eval.alchemParams(alphas);
+    }
+
+template<class evaluator>
+inline void AlchemicalPotentialPair<evaluator>::pkgFinalize(AlchemyPackage<evaluator>& pkg)
+    {
+    // for all used variables, store the net force
+    for (unsigned int i = 0; i < m_alchemy_index.getNumElements(); i++)
+        for (unsigned int j = 0; j < evaluator::num_alchemical_parameters; j++)
+            if (pkg.compute_mask[i][j])
+                {
+                if (m_normalized)
+                    m_alchemical_particles[j * m_alchemy_index.getNumElements() + i]->setNetForce(
+                        pkg.normalizationValues[i]);
+                else
+                    m_alchemical_particles[j * m_alchemy_index.getNumElements() + i]->setNetForce();
+                }
+    }
```

Comment:
This is mostly required for the way I implemented normalization (an important part of adding constraints for digital alchemy). It could likely be made a destructor method for the package instead.